File: TCPIP/src/52/c/arp_implementation.c
/*
 * ARP Protocol Implementation in C
 * Demonstrates ARP packet creation, parsing, and cache management
 * Requires root privileges and raw socket access
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <netinet/in.h>
#include <netinet/if_ether.h>
#include <net/if.h>
#include <arpa/inet.h>
#include <linux/if_packet.h>
#include <time.h>

// ARP Header Structure (28 bytes for IPv4 over Ethernet)
struct arp_header {
    uint16_t hw_type;           // Hardware type (1 = Ethernet)
    uint16_t proto_type;        // Protocol type (0x0800 = IPv4)
    uint8_t  hw_addr_len;       // Hardware address length (6 for MAC)
    uint8_t  proto_addr_len;    // Protocol address length (4 for IPv4)
    uint16_t operation;         // Operation (1 = Request, 2 = Reply)
    uint8_t  sender_hw_addr[6]; // Sender MAC address
    uint8_t  sender_proto_addr[4]; // Sender IP address
    uint8_t  target_hw_addr[6]; // Target MAC address
    uint8_t  target_proto_addr[4]; // Target IP address
} __attribute__((packed));

// Complete Ethernet frame with ARP payload
struct arp_packet {
    struct ethhdr eth_header;
    struct arp_header arp;
} __attribute__((packed));

// ARP Cache Entry
typedef struct arp_cache_entry {
    uint32_t ip_addr;
    uint8_t mac_addr[6];
    time_t timestamp;
    int is_static;
    struct arp_cache_entry *next;
} arp_cache_entry_t;

// ARP Cache
typedef struct {
    arp_cache_entry_t *head;
    int timeout_seconds;
} arp_cache_t;

// Initialize ARP cache
arp_cache_t* arp_cache_init(int timeout) {
    arp_cache_t *cache = malloc(sizeof(arp_cache_t));
    cache->head = NULL;
    cache->timeout_seconds = timeout;
    return cache;
}

// Add entry to ARP cache
void arp_cache_add(arp_cache_t *cache, uint32_t ip, uint8_t *mac, int is_static) {
    // Check if entry exists
    arp_cache_entry_t *current = cache->head;
    while (current) {
        if (current->ip_addr == ip) {
            // Update existing entry
            memcpy(current->mac_addr, mac, 6);
            current->timestamp = time(NULL);
            current->is_static = is_static;
            return;
        }
        current = current->next;
    }
    
    // Create new entry
    arp_cache_entry_t *entry = malloc(sizeof(arp_cache_entry_t));
    entry->ip_addr = ip;
    memcpy(entry->mac_addr, mac, 6);
    entry->timestamp = time(NULL);
    entry->is_static = is_static;
    entry->next = cache->head;
    cache->head = entry;
}

// Lookup MAC address in cache
int arp_cache_lookup(arp_cache_t *cache, uint32_t ip, uint8_t *mac_out) {
    time_t now = time(NULL);
    arp_cache_entry_t *current = cache->head;
    
    while (current) {
        if (current->ip_addr == ip) {
            // Check if entry is expired
            if (current->is_static || 
                (now - current->timestamp) < cache->timeout_seconds) {
                memcpy(mac_out, current->mac_addr, 6);
                return 1; // Found
            }
        }
        current = current->next;
    }
    return 0; // Not found
}

// Remove expired entries
void arp_cache_cleanup(arp_cache_t *cache) {
    time_t now = time(NULL);
    arp_cache_entry_t *current = cache->head;
    arp_cache_entry_t *prev = NULL;
    
    while (current) {
        if (!current->is_static && 
            (now - current->timestamp) >= cache->timeout_seconds) {
            // Remove expired entry
            if (prev) {
                prev->next = current->next;
            } else {
                cache->head = current->next;
            }
            arp_cache_entry_t *to_free = current;
            current = current->next;
            free(to_free);
        } else {
            prev = current;
            current = current->next;
        }
    }
}

// Print ARP cache
void arp_cache_print(arp_cache_t *cache) {
    printf("\nARP Cache Contents:\n");
    printf("%-15s %-17s %-10s\n", "IP Address", "MAC Address", "Type");
    printf("================================================\n");
    
    arp_cache_entry_t *current = cache->head;
    while (current) {
        struct in_addr ip;
        ip.s_addr = current->ip_addr;
        printf("%-15s %02x:%02x:%02x:%02x:%02x:%02x %-10s\n",
               inet_ntoa(ip),
               current->mac_addr[0], current->mac_addr[1],
               current->mac_addr[2], current->mac_addr[3],
               current->mac_addr[4], current->mac_addr[5],
               current->is_static ? "Static" : "Dynamic");
        current = current->next;
    }
}

// Create ARP request packet
void create_arp_request(struct arp_packet *packet,
                       const char *src_ip,
                       const uint8_t *src_mac,
                       const char *target_ip) {
    // Ethernet header
    memset(packet->eth_header.h_dest, 0xff, 6); // Broadcast MAC
    memcpy(packet->eth_header.h_source, src_mac, 6);
    packet->eth_header.h_proto = htons(ETH_P_ARP);
    
    // ARP header
    packet->arp.hw_type = htons(1); // Ethernet
    packet->arp.proto_type = htons(ETH_P_IP);
    packet->arp.hw_addr_len = 6;
    packet->arp.proto_addr_len = 4;
    packet->arp.operation = htons(1); // ARP Request
    
    // Sender addresses
    memcpy(packet->arp.sender_hw_addr, src_mac, 6);
    inet_pton(AF_INET, src_ip, packet->arp.sender_proto_addr);
    
    // Target addresses
    memset(packet->arp.target_hw_addr, 0, 6); // Unknown
    inet_pton(AF_INET, target_ip, packet->arp.target_proto_addr);
}

// Create ARP reply packet
void create_arp_reply(struct arp_packet *packet,
                     const char *src_ip,
                     const uint8_t *src_mac,
                     const char *target_ip,
                     const uint8_t *target_mac) {
    // Ethernet header
    memcpy(packet->eth_header.h_dest, target_mac, 6);
    memcpy(packet->eth_header.h_source, src_mac, 6);
    packet->eth_header.h_proto = htons(ETH_P_ARP);
    
    // ARP header
    packet->arp.hw_type = htons(1);
    packet->arp.proto_type = htons(ETH_P_IP);
    packet->arp.hw_addr_len = 6;
    packet->arp.proto_addr_len = 4;
    packet->arp.operation = htons(2); // ARP Reply
    
    // Sender addresses
    memcpy(packet->arp.sender_hw_addr, src_mac, 6);
    inet_pton(AF_INET, src_ip, packet->arp.sender_proto_addr);
    
    // Target addresses
    memcpy(packet->arp.target_hw_addr, target_mac, 6);
    inet_pton(AF_INET, target_ip, packet->arp.target_proto_addr);
}

// Create Gratuitous ARP packet
void create_gratuitous_arp(struct arp_packet *packet,
                          const char *ip,
                          const uint8_t *mac) {
    // Ethernet header - broadcast
    memset(packet->eth_header.h_dest, 0xff, 6);
    memcpy(packet->eth_header.h_source, mac, 6);
    packet->eth_header.h_proto = htons(ETH_P_ARP);
    
    // ARP header
    packet->arp.hw_type = htons(1);
    packet->arp.proto_type = htons(ETH_P_IP);
    packet->arp.hw_addr_len = 6;
    packet->arp.proto_addr_len = 4;
    packet->arp.operation = htons(2); // ARP Reply (some use Request)
    
    // Both sender and target have same IP
    memcpy(packet->arp.sender_hw_addr, mac, 6);
    inet_pton(AF_INET, ip, packet->arp.sender_proto_addr);
    memset(packet->arp.target_hw_addr, 0, 6);
    inet_pton(AF_INET, ip, packet->arp.target_proto_addr);
}

// Parse ARP packet
void parse_arp_packet(const struct arp_packet *packet) {
    char sender_ip[INET_ADDRSTRLEN];
    char target_ip[INET_ADDRSTRLEN];
    
    inet_ntop(AF_INET, packet->arp.sender_proto_addr, sender_ip, INET_ADDRSTRLEN);
    inet_ntop(AF_INET, packet->arp.target_proto_addr, target_ip, INET_ADDRSTRLEN);
    
    uint16_t op = ntohs(packet->arp.operation);
    
    printf("\n--- ARP Packet ---\n");
    printf("Operation: %s\n", op == 1 ? "Request" : "Reply");
    printf("Sender MAC: %02x:%02x:%02x:%02x:%02x:%02x\n",
           packet->arp.sender_hw_addr[0], packet->arp.sender_hw_addr[1],
           packet->arp.sender_hw_addr[2], packet->arp.sender_hw_addr[3],
           packet->arp.sender_hw_addr[4], packet->arp.sender_hw_addr[5]);
    printf("Sender IP: %s\n", sender_ip);
    printf("Target MAC: %02x:%02x:%02x:%02x:%02x:%02x\n",
           packet->arp.target_hw_addr[0], packet->arp.target_hw_addr[1],
           packet->arp.target_hw_addr[2], packet->arp.target_hw_addr[3],
           packet->arp.target_hw_addr[4], packet->arp.target_hw_addr[5]);
    printf("Target IP: %s\n", target_ip);
}

// Send ARP packet
int send_arp_packet(const char *interface, const struct arp_packet *packet) {
    int sockfd;
    struct ifreq ifr;
    struct sockaddr_ll addr;
    
    // Create raw socket
    sockfd = socket(AF_PACKET, SOCK_RAW, htons(ETH_P_ARP));
    if (sockfd < 0) {
        perror("socket");
        return -1;
    }
    
    // Get interface index
    strncpy(ifr.ifr_name, interface, IFNAMSIZ - 1);
    if (ioctl(sockfd, SIOCGIFINDEX, &ifr) < 0) {
        perror("ioctl SIOCGIFINDEX");
        close(sockfd);
        return -1;
    }
    
    // Set up destination address
    memset(&addr, 0, sizeof(addr));
    addr.sll_family = AF_PACKET;
    addr.sll_ifindex = ifr.ifr_ifindex;
    addr.sll_halen = 6;
    memcpy(addr.sll_addr, packet->eth_header.h_dest, 6);
    
    // Send packet
    if (sendto(sockfd, packet, sizeof(struct arp_packet), 0,
               (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        perror("sendto");
        close(sockfd);
        return -1;
    }
    
    close(sockfd);
    return 0;
}

// Get MAC address of interface
int get_mac_address(const char *interface, uint8_t *mac) {
    int sockfd;
    struct ifreq ifr;
    
    sockfd = socket(AF_INET, SOCK_DGRAM, 0);
    if (sockfd < 0) {
        perror("socket");
        return -1;
    }
    
    strncpy(ifr.ifr_name, interface, IFNAMSIZ - 1);
    if (ioctl(sockfd, SIOCGIFHWADDR, &ifr) < 0) {
        perror("ioctl SIOCGIFHWADDR");
        close(sockfd);
        return -1;
    }
    
    memcpy(mac, ifr.ifr_hwaddr.sa_data, 6);
    close(sockfd);
    return 0;
}

// Example usage
int main(int argc, char *argv[]) {
    if (geteuid() != 0) {
        fprintf(stderr, "This program requires root privileges\n");
        return 1;
    }
    
    // Example: Create ARP cache
    arp_cache_t *cache = arp_cache_init(300); // 5 minutes timeout
    
    // Add some example entries
    uint8_t mac1[] = {0x00, 0x11, 0x22, 0x33, 0x44, 0x55};
    uint8_t mac2[] = {0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff};
    
    struct in_addr ip1, ip2;
    inet_pton(AF_INET, "192.168.1.100", &ip1);
    inet_pton(AF_INET, "192.168.1.1", &ip2);
    
    arp_cache_add(cache, ip1.s_addr, mac1, 0); // Dynamic entry
    arp_cache_add(cache, ip2.s_addr, mac2, 1); // Static entry
    
    // Print cache
    arp_cache_print(cache);
    
    // Example: Create and display ARP request
    const char *interface = "eth0"; // Change to your interface
    uint8_t src_mac[6];
    
    if (get_mac_address(interface, src_mac) == 0) {
        struct arp_packet request;
        create_arp_request(&request, "192.168.1.10", src_mac, "192.168.1.1");
        
        printf("\nCreated ARP Request:\n");
        parse_arp_packet(&request);
        
        // Uncomment to actually send the packet
        // send_arp_packet(interface, &request);
        
        // Example: Create Gratuitous ARP
        struct arp_packet garp;
        create_gratuitous_arp(&garp, "192.168.1.10", src_mac);
        printf("\nCreated Gratuitous ARP:\n");
        parse_arp_packet(&garp);
    }
    
    // Cleanup
    arp_cache_cleanup(cache);
    
    return 0;
}

File: TCPIP/src/52/rust/arp_implementation.rs
/*
 * ARP Protocol Implementation in Rust
 * Demonstrates ARP packet creation, parsing, and cache management
 * Uses pnet for packet manipulation (add to Cargo.toml: pnet = "0.34")
 */

use std::collections::HashMap;
use std::net::Ipv4Addr;
use std::time::{Duration, SystemTime};

// ARP Operation codes
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u16)]
pub enum ArpOperation {
    Request = 1,
    Reply = 2,
}

impl ArpOperation {
    fn from_u16(value: u16) -> Option<Self> {
        match value {
            1 => Some(ArpOperation::Request),
            2 => Some(ArpOperation::Reply),
            _ => None,
        }
    }
}

// MAC Address type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct MacAddress([u8; 6]);

impl MacAddress {
    pub fn new(bytes: [u8; 6]) -> Self {
        MacAddress(bytes)
    }
    
    pub fn broadcast() -> Self {
        MacAddress([0xff; 6])
    }
    
    pub fn zero() -> Self {
        MacAddress([0; 6])
    }
    
    pub fn as_bytes(&self) -> &[u8; 6] {
        &self.0
    }
}

impl std::fmt::Display for MacAddress {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:02x}:{:02x}:{:02x}:{:02x}:{:02x}:{:02x}",
               self.0[0], self.0[1], self.0[2], 
               self.0[3], self.0[4], self.0[5])
    }
}

// ARP Packet structure
#[derive(Debug, Clone)]
pub struct ArpPacket {
    pub hw_type: u16,              // Hardware type (1 = Ethernet)
    pub proto_type: u16,           // Protocol type (0x0800 = IPv4)
    pub hw_addr_len: u8,           // Hardware address length
    pub proto_addr_len: u8,        // Protocol address length
    pub operation: ArpOperation,   // Operation (Request/Reply)
    pub sender_hw_addr: MacAddress,
    pub sender_proto_addr: Ipv4Addr,
    pub target_hw_addr: MacAddress,
    pub target_proto_addr: Ipv4Addr,
}

impl ArpPacket {
    // Create ARP request
    pub fn new_request(
        sender_mac: MacAddress,
        sender_ip: Ipv4Addr,
        target_ip: Ipv4Addr,
    ) -> Self {
        ArpPacket {
            hw_type: 1,
            proto_type: 0x0800,
            hw_addr_len: 6,
            proto_addr_len: 4,
            operation: ArpOperation::Request,
            sender_hw_addr: sender_mac,
            sender_proto_addr: sender_ip,
            target_hw_addr: MacAddress::zero(),
            target_proto_addr: target_ip,
        }
    }
    
    // Create ARP reply
    pub fn new_reply(
        sender_mac: MacAddress,
        sender_ip: Ipv4Addr,
        target_mac: MacAddress,
        target_ip: Ipv4Addr,
    ) -> Self {
        ArpPacket {
            hw_type: 1,
            proto_type: 0x0800,
            hw_addr_len: 6,
            proto_addr_len: 4,
            operation: ArpOperation::Reply,
            sender_hw_addr: sender_mac,
            sender_proto_addr: sender_ip,
            target_hw_addr: target_mac,
            target_proto_addr: target_ip,
        }
    }
    
    // Create Gratuitous ARP
    pub fn new_gratuitous(mac: MacAddress, ip: Ipv4Addr) -> Self {
        ArpPacket {
            hw_type: 1,
            proto_type: 0x0800,
            hw_addr_len: 6,
            proto_addr_len: 4,
            operation: ArpOperation::Reply, // or Request
            sender_hw_addr: mac,
            sender_proto_addr: ip,
            target_hw_addr: MacAddress::zero(),
            target_proto_addr: ip, // Same as sender
        }
    }
    
    // Serialize to bytes
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::with_capacity(28);
        
        bytes.extend_from_slice(&self.hw_type.to_be_bytes());
        bytes.extend_from_slice(&self.proto_type.to_be_bytes());
        bytes.push(self.hw_addr_len);
        bytes.push(self.proto_addr_len);
        bytes.extend_from_slice(&(self.operation as u16).to_be_bytes());
        bytes.extend_from_slice(self.sender_hw_addr.as_bytes());
        bytes.extend_from_slice(&self.sender_proto_addr.octets());
        bytes.extend_from_slice(self.target_hw_addr.as_bytes());
        bytes.extend_from_slice(&self.target_proto_addr.octets());
        
        bytes
    }
    
    // Deserialize from bytes
    pub fn from_bytes(data: &[u8]) -> Result<Self, String> {
        if data.len() < 28 {
            return Err("Insufficient data for ARP packet".to_string());
        }
        
        let hw_type = u16::from_be_bytes([data[0], data[1]]);
        let proto_type = u16::from_be_bytes([data[2], data[3]]);
        let hw_addr_len = data[4];
        let proto_addr_len = data[5];
        let operation = u16::from_be_bytes([data[6], data[7]]);
        
        let operation = ArpOperation::from_u16(operation)
            .ok_or("Invalid ARP operation")?;
        
        let sender_hw_addr = MacAddress::new([
            data[8], data[9], data[10], data[11], data[12], data[13]
        ]);
        
        let sender_proto_addr = Ipv4Addr::new(
            data[14], data[15], data[16], data[17]
        );
        
        let target_hw_addr = MacAddress::new([
            data[18], data[19], data[20], data[21], data[22], data[23]
        ]);
        
        let target_proto_addr = Ipv4Addr::new(
            data[24], data[25], data[26], data[27]
        );
        
        Ok(ArpPacket {
            hw_type,
            proto_type,
            hw_addr_len,
            proto_addr_len,
            operation,
            sender_hw_addr,
            sender_proto_addr,
            target_hw_addr,
            target_proto_addr,
        })
    }
    
    // Check if this is a Gratuitous ARP
    pub fn is_gratuitous(&self) -> bool {
        self.sender_proto_addr == self.target_proto_addr
    }
}

impl std::fmt::Display for ArpPacket {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "ARP {:?}\n", self.operation)?;
        write!(f, "  Sender: {} ({})\n", 
               self.sender_proto_addr, self.sender_hw_addr)?;
        write!(f, "  Target: {} ({})", 
               self.target_proto_addr, self.target_hw_addr)
    }
}

// ARP Cache Entry
#[derive(Debug, Clone)]
struct ArpCacheEntry {
    mac_addr: MacAddress,
    timestamp: SystemTime,
    is_static: bool,
}

// ARP Cache
pub struct ArpCache {
    entries: HashMap<Ipv4Addr, ArpCacheEntry>,
    timeout: Duration,
}

impl ArpCache {
    pub fn new(timeout_seconds: u64) -> Self {
        ArpCache {
            entries: HashMap::new(),
            timeout: Duration::from_secs(timeout_seconds),
        }
    }
    
    // Add or update cache entry
    pub fn add(&mut self, ip: Ipv4Addr, mac: MacAddress, is_static: bool) {
        let entry = ArpCacheEntry {
            mac_addr: mac,
            timestamp: SystemTime::now(),
            is_static,
        };
        self.entries.insert(ip, entry);
    }
    
    // Lookup MAC address
    pub fn lookup(&self, ip: &Ipv4Addr) -> Option<MacAddress> {
        self.entries.get(ip).and_then(|entry| {
            if entry.is_static {
                return Some(entry.mac_addr);
            }
            
            // Check if entry has expired
            if let Ok(elapsed) = entry.timestamp.elapsed() {
                if elapsed < self.timeout {
                    return Some(entry.mac_addr);
                }
            }
            None
        })
    }
    
    // Remove expired entries
    pub fn cleanup(&mut self) {
        let now = SystemTime::now();
        self.entries.retain(|_, entry| {
            if entry.is_static {
                return true;
            }
            
            if let Ok(elapsed) = now.duration_since(entry.timestamp) {
                elapsed < self.timeout
            } else {
                false
            }
        });
    }
    
    // Remove specific entry
    pub fn remove(&mut self, ip: &Ipv4Addr) -> bool {
        self.entries.remove(ip).is_some()
    }
    
    // Clear all entries
    pub fn clear(&mut self) {
        self.entries.clear();
    }
    
    // Get all entries (for display)
    pub fn entries(&self) -> Vec<(Ipv4Addr, MacAddress, bool)> {
        self.entries
            .iter()
            .map(|(ip, entry)| (*ip, entry.mac_addr, entry.is_static))
            .collect()
    }
    
    // Print cache contents
    pub fn print(&self) {
        println!("\nARP Cache Contents:");
        println!("{:<15} {:<17} {:<10}", "IP Address", "MAC Address", "Type");
        println!("{}", "=".repeat(50));
        
        for (ip, entry) in &self.entries {
            println!("{:<15} {:<17} {:<10}",
                     ip,
                     entry.mac_addr,
                     if entry.is_static { "Static" } else { "Dynamic" });
        }
    }
}

// ARP Handler - manages ARP operations
pub struct ArpHandler {
    cache: ArpCache,
    local_ip: Ipv4Addr,
    local_mac: MacAddress,
}

impl ArpHandler {
    pub fn new(local_ip: Ipv4Addr, local_mac: MacAddress, cache_timeout: u64) -> Self {
        ArpHandler {
            cache: ArpCache::new(cache_timeout),
            local_ip,
            local_mac,
        }
    }
    
    // Handle incoming ARP packet
    pub fn handle_packet(&mut self, packet: &ArpPacket) -> Option<ArpPacket> {
        println!("Received ARP packet:\n{}", packet);
        
        // Update cache with sender's information
        self.cache.add(
            packet.sender_proto_addr,
            packet.sender_hw_addr,
            false,
        );
        
        // Check if packet is for us
        if packet.target_proto_addr != self.local_ip {
            return None;
        }
        
        match packet.operation {
            ArpOperation::Request => {
                // Send ARP reply
                Some(ArpPacket::new_reply(
                    self.local_mac,
                    self.local_ip,
                    packet.sender_hw_addr,
                    packet.sender_proto_addr,
                ))
            }
            ArpOperation::Reply => {
                // Just update cache (already done above)
                None
            }
        }
    }
    
    // Resolve IP to MAC (would trigger ARP request in real implementation)
    pub fn resolve(&mut self, ip: Ipv4Addr) -> Option<MacAddress> {
        if let Some(mac) = self.cache.lookup(&ip) {
            return Some(mac);
        }
        
        // In real implementation, send ARP request here
        println!("Would send ARP request for {}", ip);
        None
    }
    
    // Send gratuitous ARP
    pub fn send_gratuitous(&self) -> ArpPacket {
        ArpPacket::new_gratuitous(self.local_mac, self.local_ip)
    }
    
    // Get cache reference
    pub fn cache(&self) -> &ArpCache {
        &self.cache
    }
    
    // Get mutable cache reference
    pub fn cache_mut(&mut self) -> &mut ArpCache {
        &mut self.cache
    }
}

// Example: ARP Spoof Detection
pub struct ArpSpoofDetector {
    known_mappings: HashMap<Ipv4Addr, MacAddress>,
}

impl ArpSpoofDetector {
    pub fn new() -> Self {
        ArpSpoofDetector {
            known_mappings: HashMap::new(),
        }
    }
    
    // Check if packet is potentially malicious
    pub fn check_packet(&mut self, packet: &ArpPacket) -> bool {
        let ip = packet.sender_proto_addr;
        let mac = packet.sender_hw_addr;
        
        if let Some(&known_mac) = self.known_mappings.get(&ip) {
            if known_mac != mac {
                println!("⚠️  ARP Spoof detected!");
                println!("   IP: {}", ip);
                println!("   Known MAC: {}", known_mac);
                println!("   Received MAC: {}", mac);
                return false; // Potential spoof
            }
        } else {
            // First time seeing this IP, record it
            self.known_mappings.insert(ip, mac);
        }
        
        true // OK
    }
}

// Example usage and tests
fn main() {
    println!("=== ARP Protocol Implementation Demo ===\n");
    
    // Create local MAC and IP
    let local_mac = MacAddress::new([0x00, 0x11, 0x22, 0x33, 0x44, 0x55]);
    let local_ip = "192.168.1.10".parse().unwrap();
    
    // Create ARP handler
    let mut handler = ArpHandler::new(local_ip, local_mac, 300);
    
    // Add some static entries
    let gateway_mac = MacAddress::new([0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff]);
    let gateway_ip = "192.168.1.1".parse().unwrap();
    handler.cache_mut().add(gateway_ip, gateway_mac, true);
    
    // Demo 1: ARP Request
    println!("--- Demo 1: ARP Request ---");
    let target_ip: Ipv4Addr = "192.168.1.100".parse().unwrap();
    let request = ArpPacket::new_request(local_mac, local_ip, target_ip);
    println!("{}\n", request);
    
    // Serialize and deserialize
    let bytes = request.to_bytes();
    println!("Serialized to {} bytes", bytes.len());
    let parsed = ArpPacket::from_bytes(&bytes).unwrap();
    println!("Deserialized successfully\n");
    
    // Demo 2: ARP Reply
    println!("--- Demo 2: ARP Reply ---");
    let reply = ArpPacket::new_reply(
        MacAddress::new([0x11, 0x22, 0x33, 0x44, 0x55, 0x66]),
        target_ip,
        local_mac,
        local_ip,
    );
    println!("{}\n", reply);
    
    // Demo 3: Gratuitous ARP
    println!("--- Demo 3: Gratuitous ARP ---");
    let garp = handler.send_gratuitous();
    println!("{}", garp);
    println!("Is Gratuitous: {}\n", garp.is_gratuitous());
    
    // Demo 4: Handle incoming packet
    println!("--- Demo 4: Handle Incoming ARP Request ---");
    let incoming_request = ArpPacket::new_request(
        MacAddress::new([0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb]),
        "192.168.1.50".parse().unwrap(),
        local_ip,
    );
    
    if let Some(response) = handler.handle_packet(&incoming_request) {
        println!("\nGenerated response:");
        println!("{}\n", response);
    }
    
    // Demo 5: Cache operations
    println!("--- Demo 5: ARP Cache ---");
    handler.cache().print();
    
    // Demo 6: Resolve IP
    println!("\n--- Demo 6: IP Resolution ---");
    if let Some(mac) = handler.resolve(gateway_ip) {
        println!("Resolved {} -> {}", gateway_ip, mac);
    }
    
    if handler.resolve("192.168.1.200".parse().unwrap()).is_none() {
        println!("192.168.1.200 not in cache (would send ARP request)");
    }
    
    // Demo 7: Spoof detection
    println!("\n--- Demo 7: ARP Spoof Detection ---");
    let mut detector = ArpSpoofDetector::new();
    
    let legit_packet = ArpPacket::new_reply(
        gateway_mac,
        gateway_ip,
        local_mac,
        local_ip,
    );
    detector.check_packet(&legit_packet);
    
    // Simulate spoof attempt
    let spoof_packet = ArpPacket::new_reply(
        MacAddress::new([0xff, 0xff, 0xff, 0xff, 0xff, 0xff]),
        gateway_ip, // Same IP, different MAC
        local_mac,
        local_ip,
    );
    detector.check_packet(&spoof_packet);
    
    println!("\n=== Demo Complete ===");
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_arp_packet_serialization() {
        let mac = MacAddress::new([0x00, 0x11, 0x22, 0x33, 0x44, 0x55]);
        let ip = "192.168.1.1".parse().unwrap();
        let packet = ArpPacket::new_request(mac, ip, "192.168.1.2".parse().unwrap());
        
        let bytes = packet.to_bytes();
        let parsed = ArpPacket::from_bytes(&bytes).unwrap();
        
        assert_eq!(parsed.operation, ArpOperation::Request);
        assert_eq!(parsed.sender_hw_addr, mac);
        assert_eq!(parsed.sender_proto_addr, ip);
    }
    
    #[test]
    fn test_cache_operations() {
        let mut cache = ArpCache::new(60);
        let mac = MacAddress::new([0x00, 0x11, 0x22, 0x33, 0x44, 0x55]);
        let ip = "192.168.1.1".parse().unwrap();
        
        cache.add(ip, mac, false);
        assert_eq!(cache.lookup(&ip), Some(mac));
        
        cache.remove(&ip);
        assert_eq!(cache.lookup(&ip), None);
    }
    
    #[test]
    fn test_gratuitous_arp_detection() {
        let mac = MacAddress::new([0x00, 0x11, 0x22, 0x33, 0x44, 0x55]);
        let ip = "192.168.1.1".parse().unwrap();
        let garp = ArpPacket::new_gratuitous(mac, ip);
        
        assert!(garp.is_gratuitous());
        assert_eq!(garp.sender_proto_addr, garp.target_proto_addr);
    }
}
